Code/Framework/AzCore/AzCore/Module/Environment.h

```diff
@@ -426,12 +428,16 @@ namespace AZ
             return *reinterpret_cast<T*>(&m_data->m_value);
         }
 
-        void Set(const T& value)
+        template<typename U = T>
+        AZStd::enable_if_t<AZStd::is_copy_assignable_v<U>>
+        Set(const T& value)
```

Comment:
You don't actually need to use a template here.
You can use enable_if_t with a trailing return type.
```suggestion
        auto Set(const T& value) -> AZStd::enable_if_t<AZStd::is_copy_assignable_v<U>>
```

Alternatively you can use the [AZStd::convertible_to](https://en.cppreference.com/w/cpp/concepts/convertible_to) concept and reduce this to a single function.
```suggestion
        template<typename U>
        auto Set(U&& value) -> enable_if_t<AZStd::convertible_to<U, T>>
```

Reply:
Actually I think I want just `AZStd::is_assignable`. `AZStd::is_assignable_v<T&, U>` will tell me if `Get() = forward<U>(value)` is well-formed or not.

```suggestion
template<typename U>
auto Set(U&& value) -> AZStd::enable_if_t<AZStd::is_assignable_v<T&, U>> 
{
    Get() = AZStd::forward<U>(value);
}
```